### ISSUE TITLE
Adding concurrency to query rewrite?

### ISSUE THREAD

**Issue by zhaih** [apache/lucene#11838]
### Description

There're several queries relying on rewrite to do the most heavy work, such as KnnVectorQuery and AutomatonQuery, should we make rewrite step being able to run under multi-threading? Or we could also just move those costly steps into scorer creation and don't rewrite, that could also be an option because I can't really come up with other queries that are costly when doing rewrite.

**Comment by jtibshirani** [comment]
This is a great question! To me it'd make sense to try to move costly steps into weight creation or scoring. It feels a little "off" to do the bulk of a query's work during rewrite.

However, I understand why we chose to use rewrite for `KnnVectorQuery`. It allows us the query to match the global top `k` across segments, as opposed to matching `k` docs per segment. This means the number of documents it matches doesn't depend on the number of segments, and it's easier to combine with other queries. Maybe this is not the right trade-off though in terms of behavior vs. performance 🤔

**Comment by uschindler** [comment]
For AutomatonQuery it is also required to rewrite, because there are different implementations used. When you expand to a BooleanQuery or TermInSetQuery, you need to do that in the rewrite phase.

My question is: What brings moving it to the Weight? It is still not multithreaded!

Maybe a good idea would be to allow the heavy calculations in Weight and also Rewrite to execute in an Executor. By default we pass a SameThreadExecutor implementation (basically the executor just calls `run()`), but a query implementation might use multiple threads by splitting the work into smaller pieces. E.g. a BooleanQuery could rewrite all its clauses in a separate task.

**Comment by zhaih** [comment]
> Maybe a good idea would be to allow the heavy calculations in Weight and also Rewrite to execute in an Executor.

+1, although I probably would like to start with rewrite first since I'm not aware of any weight creation that could easily benefit from execute it in parallel. 

I'm thinking about something similar to what @uschindler suggest, basically create another public method `rewrite(IndexReader, Executor)` on Query and default it to just call `rewrite(IndexReader)` without any action using executor, while in the IndexSearcher we could always pass in whatever executor we have. And then from there we could work on individual query's rewriting implementation. 

Seems people are not going against this idea I'll get a PR out soon, and we can start from there :)

**Comment by zhaih** [comment]
Change to main branch merged, keep this open until backport finishes

**Comment by benwtrent** [comment]
Merged and backported for a future Lucene release
